File: eva-core/src/main/java/io/nerv/core/advice/NoRepeatSubmitAdvice.java
package io.nerv.core.advice;

import cn.hutool.cache.Cache;
import lombok.extern.slf4j.Slf4j;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Component;
import org.springframework.web.context.request.RequestContextHolder;
import org.springframework.web.context.request.ServletRequestAttributes;

import javax.servlet.http.HttpServletRequest;

/**
 * 防止重复提交切入点
 */
@Aspect
@Component
@Slf4j
public class NoRepeatSubmitAdvice {
    @Autowired
    @Qualifier("fifoCache")
    private Cache<String, Integer> cache;

    @Around("@annotation(io.nerv.core.annotation.NoRepeatSubmit)")
    public Object arround(ProceedingJoinPoint pjp) {
        try {
            ServletRequestAttributes attributes = (ServletRequestAttributes) RequestContextHolder.getRequestAttributes();
            String sessionId = RequestContextHolder.getRequestAttributes().getSessionId();
            HttpServletRequest request = attributes.getRequest();
            String key = sessionId + "-" + request.getServletPath();
            if (cache.get(key) == null) {// 如果缓存中有这个url视为重复提交
                Object o = pjp.proceed();
                cache.put(key, 0);
                return o;
            } else {
                log.error("重复提交");
                return null;
            }
        } catch (Throwable e) {
            e.printStackTrace();
            log.error("验证重复提交时出现未知异常!");
            return "{\"code\":-889,\"message\":\"验证重复提交时出现未知异常!\"}";
        }

    }
}

File: eva-core/src/main/kotlin/io/nerv/core/advice/NoRepeatSubmitAdvice.kt
package io.nerv.core.advice

import cn.hutool.crypto.SecureUtil
import io.nerv.core.constant.CommonConstant
import io.nerv.core.enums.BizCodeEnum
import io.nerv.core.token.util.TokenUtil
import io.nerv.exception.BizException
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.cache.CacheManager
import org.springframework.stereotype.Component
import org.springframework.web.context.request.RequestContextHolder
import org.springframework.web.context.request.ServletRequestAttributes

/**
 * 防止重复提交切入点
 */
@Aspect
@Component
@ConditionalOnProperty(prefix = "eva", name = ["norepeat-check"], havingValue = "true")
class NoRepeatSubmitAdvice {
    var log = LoggerFactory.getLogger(this.javaClass)

    @Autowired
    private val cacheManager: CacheManager? = null

    @Autowired
    private val tokenUtil: TokenUtil? = null

    @Around("@annotation(io.nerv.core.annotation.NoRepeatSubmit)")
    fun arround(pjp: ProceedingJoinPoint): Any {
        val cache = cacheManager!!.getCache(CommonConstant.CACHE_REPEATSUBMIT)
        return try {
            val attributes = RequestContextHolder.getRequestAttributes() as ServletRequestAttributes
            val request = attributes.request
            // 请求类型
            val method = request.method
            var key: String? = method + ":" + tokenUtil!!.getToken(request) + "-" + request.servletPath
            key = SecureUtil.md5(key)

            // 如果缓存中有这个url视为重复提交
            if (cache[key] == null) {
                cache.put(key, 0)
            } else {
                throw BizException(BizCodeEnum.REQUEST_TOO_MORE)
            }
            pjp.proceed()
        } catch (e: Throwable) {
            log.error("验证重复提交时出现未知异常!", e)
            throw BizException(BizCodeEnum.REQUEST_TOO_MORE)
        }
    }
}